traits/src/ff/repr.rs

```diff
@@ -0,0 +1,94 @@
+pub trait Endian {
+    fn to_bytes(res: &mut [u8], el: &[u64]);
+    fn from_bytes(res: &[u8], el: &mut [u64]);
+}
+
+pub struct LE;
+pub struct BE;
+
+impl Endian for LE {
+    fn to_bytes(res: &mut [u8], el: &[u64]) {
+        el.iter().enumerate().for_each(|(i, limb)| {
+            let off = i * 8;
+            res[off..off + 8].copy_from_slice(&limb.to_le_bytes());
+        });
+    }
+
+    fn from_bytes(res: &[u8], el: &mut [u64]) {
+        el.iter_mut().enumerate().for_each(|(i, limb)| {
+            let off = i * 8;
+            *limb = u64::from_le_bytes(res[off..off + 8].try_into().unwrap());
+        });
+    }
+}
+impl Endian for BE {
+    fn to_bytes(res: &mut [u8], el: &[u64]) {
+        el.iter().rev().enumerate().for_each(|(i, limb)| {
+            let off = i * 8;
+            res[off..off + 8].copy_from_slice(&limb.to_be_bytes());
+        });
+    }
+
+    fn from_bytes(res: &[u8], el: &mut [u64]) {
+        el.iter_mut().rev().enumerate().for_each(|(i, limb)| {
+            let off = i * 8;
+            *limb = u64::from_be_bytes(res[off..off + 8].try_into().unwrap());
+        });
+    }
+}
+
+pub trait FieldRepr: Sized {
+    const SIZE: usize;
+
+    type Repr: Copy + Default + Send + Sync + 'static + AsRef<[u8]> + AsMut<[u8]>;
+
+    fn from_repr<E: Endian>(repr: &Self::Repr) -> subtle::CtOption<Self>;
+    fn to_repr<E: Endian>(&self) -> Self::Repr;
```

Comment:
Could we add `as_repr(&self) -> &Self::Repr`?  And assume that the internal representation is always little endian?  The reason why I'm suggesting this is so that we can have a method to access the little endian bytes without any copy involved so that we can operate.  Something like this:
```
let byte_le = value.as_repr().as_ref();
```
Were we're just taking a reference (no actual data copied) to the internal representation which we assume is little endian.

Reply:
up deriving `AsRef<[u8]>` as default little endian might do? @ed255

Reply:
> up deriving `AsRef<[u8]>` as default little endian might do? @ed255

Yes!  In particular having `PrimeField: AsRef<[u8]>` would be great, so that we can skip the `to_repr()` and avoid the extra copy.  This would require documenting that all implementations of PrimeField must have an internal little endian representation; and also assume that the environment is little endian.  This is required in order to return a little endian slice without copying.

Swapping endianess would not allow us to return a slice, and instead we'd need to create a new array and return it.

There's a small performance benefit of avoiding copy and using references directly.  Check this benchmark where I modified the older MSM implementation to use the field as byte reference directly.  https://github.com/privacy-scaling-explorations/halo2curves/blob/a0ca2513b38fb53804d18ec3810ec5938cdf1616/src/msm.rs#L430
```
End:     older k=18 ................................................................18.192ms
End:     older_prime k=18 ..........................................................13.845ms
End:     older k=19 ................................................................26.570ms
End:     older_prime k=19 ..........................................................24.226ms
End:     older k=20 ................................................................51.592ms
End:     older_prime k=20 ..........................................................47.078ms
End:     older k=21 ................................................................106.978ms
End:     older_prime k=21 ..........................................................97.385ms
End:     older k=22 ................................................................197.160ms
End:     older_prime k=22 ..........................................................169.085ms
```

Nevertheless this change would have really bad consequences if someone would like to run this on a BigEndian enviroment (as it would not work as expected), so it would be good to discuss this without hurrying.

Reply:
Update on this: I wasn't aware that the prime fields where kept in Montgomery representation internally.  Now I understand that it doesn't make sense to have a method that returns a reference of the little endian slice, because that representation doesn't exist internally.

Also the test I did that showed a performance improvement https://github.com/kilic/halo2traits/pull/1#discussion_r1576359860 had a flaw in the coefficient generation used in the test, they were all zeroes!

We could think about doing the MSM directly using the Montgomery representation, but then we may not get the advantage for small values.

Nevertheless I think it would still be great to have `fn to_uniform_bytes(&self, bytes: &mut [u8])` which writes the integer modulo p in little endian, see https://github.com/kilic/halo2traits/pull/1#discussion_r1574479930.